packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+  type ValueType as ValueType_,
+} from '../associated-types'
+import { ControlDefinition, serialize, type SchemaType } from '../definition'
+import { type SendMessage } from '../instance'
+import { ShapeDefinition } from '../shape/v1'
+import { ShapeV2Definition } from '../shape/v2'
+
+import { GroupControl } from './group-control'
+
+type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
+type KeyDefinitions = Record<string, ItemDefinition>
+
+export type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
+  readonly label?: string
+  readonly preferredLayout?:
+    | typeof Definition.Layout.Inline
+    | typeof Definition.Layout.Popover
+  readonly props: Defs
+}
+
+type ShapeData<C extends Config> = DataType_<
+  ShapeDefinition<{ type: C['props'] }>
+>
+
+type PropsData<C extends Config> = {
+  [K in keyof C['props']]?: DataType_<C['props'][K]>
+}
+
+type VersionedData<C extends Config> = {
+  [ControlDataTypeKey]:
+    | typeof Definition.v1DataType
+    | typeof ShapeV2Definition.v1DataType
+  value: PropsData<C>
+}
+
+type DataType<C extends Config> = ShapeData<C> | VersionedData<C>
+
+type ValueType<C extends Config> = {
+  [K in keyof C['props']]?: ValueType_<C['props'][K]>
+}
+
+type ResolvedValueType<C extends Config> = {
+  [K in keyof C['props']]: ResolvedValueType_<C['props'][K]>
+}
+type InstanceType<C extends Config> = GroupControl<Definition<C>>
+
+class Definition<C extends Config> extends ControlDefinition<
+  typeof Definition.type,
+  C,
+  DataType<C>,
+  ValueType<C>,
+  ResolvedValueType<C>,
+  InstanceType<C>
+> {
+  static readonly type = 'makeswift::controls::group' as const
+  static readonly v1DataType = 'group::v1' as const
+
+  static readonly Layout = {
+    Inline: `${this.type}::layout::inline`,
+    Popover: `${this.type}::layout::popover`,
+  } as const
+
+  static deserialize(
+    data: DeserializedRecord,
+    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
+  ): GroupDefinition {
+    if (data.type !== Definition.type)
+      throw new Error(
+        `Group: expected '${Definition.type}', got '${data.type}'`,
+      )
+
+    const {
+      config: { props, ...config },
+    } = z
+      .object({
+        type: z.literal(Definition.type),
+        config: z.object({
+          label: z.string().optional(),
+          preferredLayout: z
+            .union([
+              z.literal(Definition.Layout.Inline),
+              z.literal(Definition.Layout.Popover),
+            ])
+            .optional(),
+          props: z.record(z.string(), SerializationSchema.deserializedRecord),
+        }),
+      })
+      .parse(data)
+
+    const deserializedProps = mapValues(props, (itemDef) => {
+      return deserializeCallback(itemDef)
+    })
+
+    return Group({
+      props: deserializedProps,
+      ...config,
+    })
+  }
+
+  constructor(
+    config: C,
+    readonly dataType:
+      | typeof Definition.v1DataType
+      | typeof ShapeV2Definition.v1DataType,
```

Comment:
Accept data type at runtime. When constructed with `ShapeV2Definition.v1DataType`, this effectively allows `Group` functions as `ShapeV2` and makes it unnecessary to keep two copies of essentially the same class. 

We can carry this even further and fold the legacy `Shape` into this as well, but I didn't want to go there just yet.